### PR TITLE
Train config

### PR DESCRIPTION
**Pull request by midokura-silvia**
This pull request adds to the modifications from @mari-linhares.
It enables us to send messages to a worker that trigger a call to a method of a Worker instance.

This is especially interesting for Websocket workers where two Worker instances work together (local + remote) to execute commands.

I added implementations to count and list the objects at the remote workers. 
If we add the run_training function to the WebSocketServerWorker it could be executed in the same way.

### CODE REVIEW

File: syft/workers/base.py [hunk 1]
```diff
@@ -117,6 +119,21 @@ def __init__(
                         if self.id not in worker._known_workers:
                             worker.add_worker(self)
 
+    def remove_worker_from_registry(self, worker_id):
+        """remove a worker from the dictionary of known workers
```
Comment (mari-linhares) [inline syft/workers/base.py]: remove a worker from the dictionary of known workers -> Remove a worker from the dictionary of known workers.

File: syft/workers/base.py [hunk 2]
```diff
@@ -117,6 +119,21 @@ def __init__(
...
+
+        :param worker_id: id to be removed
```
Comment (mari-linhares) [inline syft/workers/base.py]: change these to [google docstring style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: syft/workers/base.py [hunk 3]
```diff
@@ -117,6 +119,21 @@ def __init__(
...
+        :return:
+        """
+        del self._known_workers[worker_id]
+
+    def remove_worker_from_local_worker_registry(self):
+        """removes itself from the registry of the hook.local_worker
+
+        :return:
```
Comment (mari-linhares) [inline syft/workers/base.py]: Same about google doscstring syle.

File: syft/workers/base.py [hunk 4]
```diff
@@ -613,7 +633,7 @@ def __str__(self):
         out = "<"
         out += str(type(self)).split("'")[1].split(".")[-1]
         out += " id:" + str(self.id)
-        out += " #tensors:" + str(len(self._objects))
+        out += " #objects:" + str(len(self._objects))
```
Comment (mari-linhares) [inline syft/workers/base.py]: Nice catch!

File: syft/workers/base.py [hunk 5]
```diff
@@ -779,3 +799,21 @@ def generate_triple(
         b_shared = b.share(*locations, field=field, crypto_provider=self).child
         c_shared = c.share(*locations, field=field, crypto_provider=self).child
         return a_shared, b_shared, c_shared
+
+    @staticmethod
+    def create_message_execute_command(
+        command_name: codes.MSGTYPE, command_owner=None, return_ids=None, *args, **kwargs
+    ):
+        """ helper function creating a message tuple for the execute_command call
+        :param command_name: name of the command that shall be called
```
Comment (mari-linhares) [inline syft/workers/base.py]: Same about google style docstrings.

File: syft/workers/websocket_client.py [hunk 6]
```diff
@@ -96,3 +94,40 @@ def _recv_msg(self, message: bin) -> bin:
...
+        serialized_message = sy.serde.serialize(message)
+        # Send the message and return the deserialized response.
+
+        response = self._recv_msg(serialized_message)
+        return sy.serde.deserialize(response)
+
+    def __str__(self):
+        """Returns the string representation of a Websocket worker.
+
+        A to-string method for websocket workers that includes information from the websocket server
```
Comment (mari-linhares) [inline syft/workers/websocket_client.py]: Missing period.

File: syft/workers/websocket_client.py [hunk 7]
```diff
@@ -96,3 +94,40 @@ def _recv_msg(self, message: bin) -> bin:
...
+
+        Returns:
+            The Type and ID of the worker
```
Comment (mari-linhares) [inline syft/workers/websocket_client.py]: Missing period.

File: syft/workers/websocket_client.py [hunk 8]
```diff
@@ -96,3 +94,40 @@ def _recv_msg(self, message: bin) -> bin:
...
+
+        """
+        out = "<"
+        out += str(type(self)).split("'")[1].split(".")[-1]
```
Comment (mari-linhares) [inline syft/workers/websocket_client.py]: This is a bit confusing, can you add a comment about what's going on on this line?

File: test/workers/test_websocket_worker.py [hunk 9]
```diff
@@ -41,17 +41,77 @@ def test_websocket_workers_search(hook, start_proc):
...
+
+    y = torch.tensor([4, 5, 6]).send(local_fed1)
+    res = local_fed1.list_objects_remote()
+    res_dict = eval(res.replace("tensor", "torch.tensor"))
+    assert len(res_dict) == 2
+
+    # delete x before terminating the websocket connection
+    del x
+    del y
+    time.sleep(1)
```
Comment (mari-linhares) [inline test/workers/test_websocket_worker.py]: Can you decrease this value (maybe 0.1)?

File: test/workers/test_websocket_worker.py [hunk 10]
```diff
@@ -41,17 +41,77 @@ def test_websocket_workers_search(hook, start_proc):
...
+    local_worker.remove_worker_from_local_worker_registry()
+    process_remote_worker.terminate()
+
+
+def test_list_objects_remote(hook, start_proc):
+
+    kwargs = {"id": "fed", "host": "localhost", "port": 8765, "hook": hook}
+    process_remote_fed1 = start_proc(WebsocketServerWorker, kwargs)
+
+    time.sleep(1)
```
Comment (mari-linhares) [inline test/workers/test_websocket_worker.py]: Can you decrease this value (maybe 0.1)?

File: test/workers/test_websocket_worker.py [hunk 11]
```diff
@@ -41,17 +41,77 @@ def test_websocket_workers_search(hook, start_proc):
...
+
+    kwargs = {"id": "fed", "host": "localhost", "port": 8765, "hook": hook}
+    local_fed1 = WebsocketClientWorker(**kwargs)
+
+    x = torch.tensor([1, 2, 3]).send(local_fed1)
+
+    res = local_fed1.list_objects_remote()
+    res_dict = eval(res.replace("tensor", "torch.tensor"))
```
Comment (mari-linhares) [inline test/workers/test_websocket_worker.py]: Why do you need this eval?

File: test/workers/test_websocket_worker.py [hunk 12]
```diff
@@ -41,17 +41,77 @@ def test_websocket_workers_search(hook, start_proc):
...
+    nr_objects = local_worker.objects_count_remote()
+    assert nr_objects == 2
+
+    x.get()
+    nr_objects = local_worker.objects_count_remote()
+    assert nr_objects == 1
+
+    # delete remote object before terminating the websocket connection
+    del y
+    time.sleep(1)
```
Comment (mari-linhares) [inline test/workers/test_websocket_worker.py]: Same.

File: test/workers/test_websocket_worker.py [hunk 13]
```diff
@@ -41,17 +41,77 @@ def test_websocket_workers_search(hook, start_proc):
...
+    local_fed1.remove_worker_from_local_worker_registry()
+    process_remote_fed1.terminate()
+
+
+def test_objects_count_remote(hook, start_proc):
+
+    kwargs = {"id": "fed", "host": "localhost", "port": 8765, "hook": hook}
+    process_remote_worker = start_proc(WebsocketServerWorker, kwargs)
+
+    time.sleep(1)
```
Comment (mari-linhares) [inline test/workers/test_websocket_worker.py]: Same.

File: syft/workers/base.py [hunk 14]
```diff
@@ -323,11 +340,14 @@ def execute_command(self, message: tuple) -> PointerTensor:
         command_name = command_name
         # Handle methods
         if _self is not None:
-            if sy.torch.is_inplace_method(command_name):
-                getattr(_self, command_name)(*args, **kwargs)
-                return
+            if type(_self) == str and _self == "self":
```
Comment (mari-linhares) [inline syft/workers/base.py]: Is this related to the main change of your PR? Maybe add a comment?

File: test/workers/test_websocket_worker.py (hunk 11 above)
Comment (midokura-silvia) [reply test/workers/test_websocket_worker.py]: Because currently the result is a string. And eval transforms it into a dictionary. 
Currently this function is only used in the __str__ method of the websocket client worker. And there a string is used.
